src/sql/src/session/vars2.rs

```diff
@@ -0,0 +1,685 @@
+            feature_flag: None,
+        }
+    }
+
+    pub fn new_runtime<V: Value>(
+        name: &'static str,
+        value: V,
+        description: &'static str,
+        internal: bool,
+    ) -> Self {
+        VarDefinition {
+            name: UncasedStr::new(name),
+            description,
+            value: VarDefaultValue::Runtime(Arc::new(value)),
+            internal,
+            parse: V::parse_dyn_value,
+            type_name: V::type_name,
+            constraint: None,
+            feature_flag: None,
+        }
+    }
+
+    /// TODO(parkmycar): Refactor this method onto a `VarDefinitionBuilder` that would allow us to
+    /// constrain `V` here to be the same `V` used in [`VarDefinition::new`].
+    pub const fn with_constraint<V: Value, S, D: DomainConstraint<V, S>>(
+        mut self,
+        _constraint: D,
+    ) -> Self {
+        self.constraint = Some(ValueConstraint::Domain(D::check_box_dyn));
+        self
+    }
+
+    pub const fn read_only(mut self) -> Self {
+        self.constraint = Some(ValueConstraint::ReadOnly);
+        self
+    }
+
+    pub const fn with_feature_flag(mut self, feature_flag: &'static FeatureFlag) -> Self {
+        self.feature_flag = Some(feature_flag);
+        self
+    }
+
+    pub fn parse(&self, input: VarInput) -> Result<Box<dyn Value>, VarError> {
+        (self.parse)(input).map_err(|kind| {
+            let value = self.value.value().format();
+            VarError {
+                name: self.name.as_str(),
+                value,
+                type_name: (self.type_name)(),
+                kind,
+            }
+        })
+    }
+}
+
+impl Var for VarDefinition {
+    fn name(&self) -> &'static str {
+        self.name.as_str()
+    }
+
+    fn value(&self) -> String {
+        self.value.value().format()
+    }
+
+    fn description(&self) -> &'static str {
+        self.description
+    }
+
+    fn type_name(&self) -> String {
+        (self.type_name)().to_string()
+    }
+
+    fn visible(
+        &self,
+        user: &User,
+        system_vars: Option<&super::vars::SystemVars>,
+    ) -> Result<(), super::vars::VarError> {
+        if self.internal && user != &*SYSTEM_USER && user != &*SUPPORT_USER {
+            todo!()
+            // Err(VarErrorKind::UnknownParameter)
+        } else if self.name().starts_with("unsafe")
+            && match system_vars {
+                None => true,
+                Some(system_vars) => !system_vars.allow_unsafe(),
+            }
+        {
+            todo!()
+            // Err(VarErrorKind::RequiresUnsafeMode)
+        } else {
+            Ok(())
+        }
+    }
+}
+
+/// The kinds of compiled in default values that can be used with [`VarDefinition`].
+#[derive(Clone, Debug)]
+pub enum VarDefaultValue {
+    /// Static that can be evaluated at compile time.
+    Static(&'static dyn Value),
+    /// Lazy value that is defined at compile time, but created at runtime.
+    Lazy(fn() -> &'static dyn Value),
```

Comment:
What would an example of this be?

Reply:
A `Vec<Ident>`, you can't create a `Vec` at compile time so it needs to be lazily instantiated.